extmod/modssl_mbedtls.c

```diff
@@ -138,6 +180,70 @@ STATIC NORETURN void mbedtls_raise_error(int err) {
     #endif
 }
 
+#ifdef MICROPY_SSL_MBEDTLS_EXTRAS
+STATIC void ssl_handle_handshake_error(mp_obj_ssl_socket_t *sslsock, int *errcode) {
+
+    uint8_t alert = MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE;
+    uint32_t flags = 0;
+    uint ret = 0;
+    if (*errcode == MBEDTLS_ERR_X509_CERT_VERIFY_FAILED) {
+
+        // The certificate may have been rejected for several reasons.
+        flags = mbedtls_ssl_get_verify_result(&sslsock->ssl);
+
+    } else if (*errcode == MBEDTLS_ERR_SSL_NO_CLIENT_CERTIFICATE) {
+        /* alert = MBEDTLS_SSL_ALERT_MSG_CERT_REQUIRED; tlsv1.3 */
+
+        mbedtls_ssl_send_alert_message(&sslsock->ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL,
+            alert);
+    }
+
+    sslsock->sock = MP_OBJ_NULL;
+    mbedtls_ssl_free(&sslsock->ssl);
+
+    if (*errcode == MBEDTLS_ERR_X509_CERT_VERIFY_FAILED) {
+        char xcbuf[512];
+        ret = mbedtls_x509_crt_verify_info(xcbuf, sizeof(xcbuf), "\n", flags);
+        // The length of the string written (not including the terminated nul byte),
+        // or a negative err code.
+        if (ret > 0) {
+            mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("%s"), xcbuf);
+        }
+    }
+
+    mbedtls_raise_error(*errcode);
+}
+#endif
+
+STATIC void ssl_async_handshake(mp_obj_ssl_socket_t *sslsock, int *errcode) {
+
+    if (
+        #if MBEDTLS_VERSION_NUMBER >= 0x03000000
+        (*errcode < 0) & (mbedtls_ssl_is_handshake_over(&sslsock->ssl) == 0) & (*errcode != MBEDTLS_ERR_SSL_CONN_EOF)
+        #else
+        (*errcode < 0) & (*errcode != MBEDTLS_ERR_SSL_CONN_EOF)
+        #endif
+        ) {
+        // Async handshake is done by mbdetls_ssl_read/write
+        // if return code is MBEDTLS_ERR_XX (i.e < 0) and handshake is not done due to
+        // handshake failure notify peer
+        // with proper error code and raise mp error with mbedtls_raise_error
+        #ifdef MICROPY_SSL_MBEDTLS_EXTRAS
+        ssl_handle_handshake_error(sslsock, errcode);
```

Comment:
It looks like this function can be merged here directly, because it's only called from this location.

Then, common code can be factored out with the #else block below.  Eg the call to `mbedtls_ssl_send_alert_message()`, and maybe the `mbedtls_ssl_free()` and the final `mbedtls_raise_error()`.